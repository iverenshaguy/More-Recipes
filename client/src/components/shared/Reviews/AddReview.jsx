import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Form from '../../shared/Forms';
import './AddReview.scss';

/**
 * @exports
 * Returns Review Wrapper
 * @param {object} props
 * @returns {JSX} RecipeWrapper
 */
const AddReview = (props) => {
  const meta = {
    title: 'Review this Recipe',
    btnText: 'Add Your Review'
  };

  const reviewFormClass = props.showReviewForm ? 'my-3' : 'd-none';

  return (
    <div className={reviewFormClass} id="review-wrapper">
      <div id="review-div">
        <div className="card card-block px-md-5 px-sm-3 py-3">
          <Form {...props} type="review" id={props.id} meta={meta} />
        </div>
      </div>
    </div>
  );
};

AddReview.propTypes = {
  id: PropTypes.number.isRequired,
  showReviewForm: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  showReviewForm: state.ui.reviewForm.isOpen,
  id: state.singleRecipe.recipe.item.recipeItem.id,
  submitError: state.singleRecipe.recipeReviews.error,
  submitting: state.singleRecipe.recipeReviews.reviewing
});

export { AddReview as AddReviewComponent };

export default connect(mapStateToProps)(AddReview);
