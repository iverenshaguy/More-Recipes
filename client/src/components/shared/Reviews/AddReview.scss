#review-wrapper {
  h4,
  hr {
    display: none;
  }
}
