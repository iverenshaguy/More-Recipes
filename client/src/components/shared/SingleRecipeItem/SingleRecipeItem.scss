@import '../../../assets/scss/modules';
@import '~bootstrap/scss/bootstrap';

#recipe-wrapper {
  .image-container {
    position: relative;

    .overlay {
      position: absolute;
      font-weight: 400;
      font-size: 2em;
      top: -5%;
      right: 2%;
      cursor: pointer;

      a:not(.dropdown-item) {
        color: white;
        text-decoration: none;
      }
    }

    @include media-breakpoint-up(md) {
      .overlay {
        font-size: 3em;
        top: -6%;
        right: 3%;
      }
    }

    @include media-breakpoint-up(lg) {
      .overlay {
        font-size: 3.7em;
        top: -8%;
        right: 3%;
      }
    }
  }
}