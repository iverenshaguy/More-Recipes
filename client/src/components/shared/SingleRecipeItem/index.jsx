import React from 'react';
import PropTypes from 'prop-types';
import { UncontrolledDropdown, DropdownToggle, DropdownMenu } from 'reactstrap';
import Reviews, { AddReview } from '../Reviews';
import { RenderIcon, RenderVoteIcon } from '../Icons';
import RecipeDetailsSection from '../RecipeDetailsSection';
import { recipeObjectPropTypes, userPropTypes } from '../../../helpers/proptypes';
import './SingleRecipeItem.scss';

const altImage = '/images/jollof-rice-img.jpg';

/**
 * @exports
 * @function SingleRecipeItem
 * @param {object} props
 * @returns {JSX} SingleRecipeItem
 */
const SingleRecipeItem = (props) => {
  const {
    recipe, user, reviewed, voteRecipe, toggleReviewForm, addRecipeToFavorites
  } = props;
  let disableRecipeActions = false;

  if (recipe.recipeItem.userId === user.id) {
    disableRecipeActions = true;
  }

  return (
    <div className="container page-margin" id="recipe-wrapper">
      <div className="recipe-picture-wrapper">
        <div className="image-container">
          <img src={recipe.recipeItem.recipeImage ? recipe.recipeItem.recipeImage : altImage} alt="recipe" className="img-fluid" />
          {/* <div className="overlay"><span>...</span></div> */}
          <UncontrolledDropdown className="overlay">
            <DropdownToggle tag="a">...</DropdownToggle>
            <DropdownMenu right>
              <a className="dropdown-item" href="#edit">Edit Recipe</a>
              <a className="dropdown-item" href="#delete">Delete Recipe</a>
            </DropdownMenu>
          </UncontrolledDropdown>
        </div>
      </div>
      <div className="row px-xs-1 px-sm-2 px-md-2 px-lg-4 px-xl-5" id="recipe-details">
        <div className="col-12 px-xl-5 px-lg-5 px-md-5 px-sm-4 px-xs-3">
          <div className="heading p-5 text-center">
            <h3>{recipe.recipeItem.recipeName}</h3>
          </div>
          <div className="row" id="recipe-rating">
            <div className="col-xs-12 col-md-6 col-lg-4 pt-3 pb-1">
              {!disableRecipeActions && <p className="user-info"><span className="text-muted">Posted By </span>{recipe.recipeItem.User.username}</p>}
            </div>
            <div className="col-xs-12 col-md-6 col-lg-8 pt-3 pb-1" id="recipe-rating-info">
              <p className="text-muted inline-p"><i className="aria-hidden fa fa-star" /> {recipe.recipeItem.rating ? (+recipe.recipeItem.rating).toPrecision(2) : 0} &nbsp; &nbsp;</p>
              <p className="text-muted inline-p"><i className="aria-hidden fa fa-heart" /> {recipe.recipeItem.upvotes} &nbsp;&nbsp;</p>
              <p className="text-muted inline-p"><i className="text-success aria-hidden fa fa-eye" /> {recipe.recipeItem.views} &nbsp;&nbsp;</p>
              <p className="text-muted inline-p"><i className="aria-hidden fa fa-thumbs-down" /> {recipe.recipeItem.downvotes} &nbsp; &nbsp;</p>
            </div>
          </div>
          <RecipeDetailsSection recipe={recipe.recipeItem} />
          {!disableRecipeActions &&
            <div className="d-flex pt-5 pb-1 align-items-center" id="call-to-action">
              <RenderIcon recipe={recipe} reviewed={reviewed} type="review" icon="star" handleClick={toggleReviewForm} />
              <RenderIcon recipe={recipe} type="favorite" icon="heart" handleClick={addRecipeToFavorites} />
              <RenderVoteIcon recipe={recipe} type="up" boolCheck handleClick={voteRecipe} />
              <RenderVoteIcon recipe={recipe} type="down" boolCheck={false} handleClick={voteRecipe} />
            </div>}
          {!disableRecipeActions &&
            !recipe.recipeItem.isReviewed &&
            !reviewed &&
            <AddReview id={recipe.recipeItem.id} />}
          <Reviews recipeId={recipe.recipeItem.id} />
        </div>
      </div>
    </div>
  );
};

SingleRecipeItem.propTypes = {
  reviewed: PropTypes.bool.isRequired,
  error: PropTypes.shape({
    recipeId: PropTypes.shape({
      msg: PropTypes.string
    })
  }),
  voteRecipe: PropTypes.func.isRequired,
  toggleReviewForm: PropTypes.func.isRequired,
  addRecipeToFavorites: PropTypes.func.isRequired,
  ...userPropTypes,
  ...recipeObjectPropTypes
};

SingleRecipeItem.defaultProps = {
  error: null
};

export default SingleRecipeItem;
