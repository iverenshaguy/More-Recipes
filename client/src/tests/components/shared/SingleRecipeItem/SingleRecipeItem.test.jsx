import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import SingleRecipeItem from '../../../../components/shared/SingleRecipeItem';
import AddReview from '../../../../components/shared/Reviews/AddReview';
import initialValues from '../../../setup/initialValues';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore(initialValues);
// const voteRecipeMock = jest.fn();
// const toggleReviewFormMock = jest.fn();
// const addRecipeToFavoritesMock = jest.fn();

const setup = () => {
  const props = {
    dispatch: jest.fn(),
    recipe: {
      recipeItem: {
        id: 2,
        recipeName: 'Recipe',
        userId: 1,
        isReviewed: false,
        ingredients: ['rice', 'beans'],
        preparations: ['wash', 'cook'],
        directions: ['wash', 'cook'],
        User: {
          id: 1,
          username: 'name'
        }
      }
    },
    error: null,
    user: {
      id: 2,
      username: 'username'
    },
    reviewed: false,
    voteRecipe: jest.fn(),
    toggleReviewForm: jest.fn(),
    addRecipeToFavorites: jest.fn(),
  };

  const shallowWrapper = shallow(<SingleRecipeItem {...props} />);
  const mountedWrapper = mount(//eslint-disable-line
    <Provider store={store}>
      <SingleRecipeItem {...props} />
    </Provider>);

  return { props, shallowWrapper, mountedWrapper };
};

describe('SingleRecipeItem', () => {
  it('renders correctly', () => {
    const { shallowWrapper } = setup();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('disables recipe username and recipe actions if recipe belongs to user', () => {
    const { props } = setup();
    const user = {
      id: 1,
      username: 'username'
    };
    const shallowWrapper = shallow(<SingleRecipeItem {...props} user={user} />);

    expect(shallowWrapper.find('p.user-info').length).toBeFalsy();
    expect(shallowWrapper.find('div#call-to-action').length).toBeFalsy();
  });

  it('shows review form when recipe is not reviewed and show review form is true', () => {
    const { props } = setup();
    const shallowWrapper = shallow(<SingleRecipeItem {...props} showReviewForm />);
    expect(shallowWrapper.find(AddReview).length).toBeTruthy();
  });

  it('calls toggleReviewForm() when review icon is clicked', () => {
    const { props, mountedWrapper } = setup();
    const wrapper = mountedWrapper.find(SingleRecipeItem);

    wrapper.find('i.fa-star-o').simulate('click');

    expect(props.toggleReviewForm).toHaveBeenCalled();
    mountedWrapper.unmount();
  });

  it('calls addRecipeToFavorites() when favorite icon is clicked', () => {
    const { props, mountedWrapper } = setup();
    const wrapper = mountedWrapper.find(SingleRecipeItem);

    wrapper.find('i.fa-heart-o').simulate('click');

    expect(props.addRecipeToFavorites).toHaveBeenCalled();
    mountedWrapper.unmount();
  });

  it('calls voteRecipe() when upvote icon is clicked', () => {
    const { props, mountedWrapper } = setup();
    const wrapper = mountedWrapper.find(SingleRecipeItem);

    wrapper.find('i.fa-thumbs-up').simulate('click');

    expect(props.voteRecipe).toHaveBeenCalled();
    mountedWrapper.unmount();
  });
});
