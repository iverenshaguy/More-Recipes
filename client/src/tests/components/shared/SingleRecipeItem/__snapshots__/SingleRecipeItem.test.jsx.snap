// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleRecipeItem renders correctly 1`] = `
<div
  className="container page-margin"
  id="recipe-wrapper"
>
  <div
    className="recipe-picture-wrapper"
  >
    <div
      className="image-container"
    >
      <img
        alt="recipe"
        className="img-fluid"
      />
    </div>
  </div>
  <div
    className="row px-xs-1 px-sm-2 px-md-2 px-lg-4 px-xl-5"
    id="recipe-details"
  >
    <div
      className="col-12 px-xl-5 px-lg-5 px-md-5 px-sm-4 px-xs-3"
    >
      <div
        className="heading p-5 text-center"
      >
        <h3>
          Recipe
        </h3>
      </div>
      <div
        className="row"
        id="recipe-rating"
      >
        <div
          className="col-xs-12 col-md-6 col-lg-4 pt-3 pb-1"
        >
          <p
            className="user-info"
          >
            <span
              className="text-muted"
            >
              Posted By 
            </span>
            name
          </p>
        </div>
        <div
          className="col-xs-12 col-md-6 col-lg-8 pt-3 pb-1"
          id="recipe-rating-info"
        >
          <p
            className="text-muted inline-p"
          >
            <i
              className="aria-hidden fa fa-star"
            />
             
            NaN
                
          </p>
          <p
            className="text-muted inline-p"
          >
            <i
              className="aria-hidden fa fa-heart"
            />
             
               
          </p>
          <p
            className="text-muted inline-p"
          >
            <i
              className="text-success aria-hidden fa fa-eye"
            />
             
               
          </p>
          <p
            className="text-muted inline-p"
          >
            <i
              className="aria-hidden fa fa-thumbs-down"
            />
             
                
          </p>
        </div>
      </div>
      <RecipeDetailsSection
        recipe={
          Object {
            "User": Object {
              "id": 1,
              "username": "name",
            },
            "directions": Array [
              "wash",
              "cook",
            ],
            "id": 2,
            "ingredients": Array [
              "rice",
              "beans",
            ],
            "isReviewed": false,
            "preparations": Array [
              "wash",
              "cook",
            ],
            "recipeName": "Recipe",
            "userId": 1,
          }
        }
      />
      <div
        className="d-flex pt-5 pb-1 align-items-center"
        id="call-to-action"
      >
        <RenderIcon
          handleClick={[Function]}
          icon="star"
          recipe={
            Object {
              "recipeItem": Object {
                "User": Object {
                  "id": 1,
                  "username": "name",
                },
                "directions": Array [
                  "wash",
                  "cook",
                ],
                "id": 2,
                "ingredients": Array [
                  "rice",
                  "beans",
                ],
                "isReviewed": false,
                "preparations": Array [
                  "wash",
                  "cook",
                ],
                "recipeName": "Recipe",
                "userId": 1,
              },
            }
          }
          reviewed={false}
          type="review"
        />
        <RenderIcon
          handleClick={[Function]}
          icon="heart"
          recipe={
            Object {
              "recipeItem": Object {
                "User": Object {
                  "id": 1,
                  "username": "name",
                },
                "directions": Array [
                  "wash",
                  "cook",
                ],
                "id": 2,
                "ingredients": Array [
                  "rice",
                  "beans",
                ],
                "isReviewed": false,
                "preparations": Array [
                  "wash",
                  "cook",
                ],
                "recipeName": "Recipe",
                "userId": 1,
              },
            }
          }
          reviewed={false}
          type="favorite"
        />
        <RenderVoteIcon
          boolCheck={true}
          handleClick={[Function]}
          recipe={
            Object {
              "recipeItem": Object {
                "User": Object {
                  "id": 1,
                  "username": "name",
                },
                "directions": Array [
                  "wash",
                  "cook",
                ],
                "id": 2,
                "ingredients": Array [
                  "rice",
                  "beans",
                ],
                "isReviewed": false,
                "preparations": Array [
                  "wash",
                  "cook",
                ],
                "recipeName": "Recipe",
                "userId": 1,
              },
            }
          }
          type="up"
        />
        <RenderVoteIcon
          boolCheck={false}
          handleClick={[Function]}
          recipe={
            Object {
              "recipeItem": Object {
                "User": Object {
                  "id": 1,
                  "username": "name",
                },
                "directions": Array [
                  "wash",
                  "cook",
                ],
                "id": 2,
                "ingredients": Array [
                  "rice",
                  "beans",
                ],
                "isReviewed": false,
                "preparations": Array [
                  "wash",
                  "cook",
                ],
                "recipeName": "Recipe",
                "userId": 1,
              },
            }
          }
          type="down"
        />
      </div>
      <Connect(AddReview)
        id={2}
      />
      <Connect(Reviews)
        recipeId={2}
      />
    </div>
  </div>
</div>
`;
